brainstorming.md

```diff
@@ -0,0 +1,372 @@
+soco.close()
+soco.uns -> metadata as a SOMAMapping
+```
+
+> ⚠️ **Issue** - do we need a specific open/close API, or is the object stateless? Or is that language/environment-specific?
+
+Collection operations include access to the collection using the SOMAMaping interface:
+
+```
+type ValueType = SOMACollection | SOMADataFrame | SOMAMatrix
+
+soco.get(string key) -> ValueType
+soco.has(string key) -> bool
+soco.set(string key, ValueType value)
+soco.del(string key) -> void
+soco.__iterator__ -> key/value iterator for collection
+soco.__length__ -> int
+```
+
+## SOMADataFrame
+
+> 🛑 **To be further specified** -- methods need additional specification.
+
+In the following, `sdf` is used to indicate an "instance" or "handle" to a SOMADataFrame.
+
+Summary:
+
+```
+SOMADataFrame.create(string uri, Arrow.Schema schema, string index_column_name) -> void
+
+SOMADataFrame.delete(string uri) -> void
+SOMADataFrame.exists(string uri) -> bool  # exists and is a SOMADataFrame
+SOMADataFrame.open(string uri, mode=read|write) -> sdf
+sdf.close()
+
+sdf.shape -> (int, )                      # length of each index dimension, always a list of length 1
+sdf.ndims -> 1                            # returns a constant
+sdf.schema -> Arrow.Schema                # data schema, which must be the schema for an Arrow.RecordBatch
+sdf.index_column_names -> (string, )      # list of index column names, always a list of length 1
+
+sdf.read(ids=[...]|all, column_names=[...]|all, filter=_TBD_) -> Arrow.RecordBatch
+sdf.write(Arrow.RecordBatch values)          # write
+```
+
+### Method: create()
+
+Create a new SOMADataFrame with user-specified URI and schema. If the schema includes unsupported types, an error will be raised.
+
+```
+SOMADataFrame.create(string uri, Arrow.Schema schema, string index_column_name) -> void
+```
+
+Parameters:
+
+- uri - location at which to create the object
+- schema - an Arrow.RecordBatch type defining all column schema, including the index columns
+- index_column_name - the name of the index column (eg, 'obs_id')
+
+### Method: read()
+
+Read a user-defined subset of data and return as an Arrow.RecordBatch. User-provided parameters specify the index values (`ids`), which columns to read, and column filters to apply. There is no guarantee of result ordering.
+
+```
+sdf.read(ids=[`string`, ...]|all, column_names=[`string`, ...]|all, ValueFilter filter=_TBD_) -> Arrow.RecordBatch
+```
+
+Parameters:
+
+- ids - list of indices to read. Defaults to 'all'
+- column_names - the named columns to read and return. Defaults to all.
+- filter - [value filter](#value-filters) to apply (_TBD_). Defaults to no filter.
+
+> ⚠️ **Issue** - sdf.read() needs further definition for:
+>
+> - incremental/streaming queries - critical gap
+> - result ordering - would be useful, but also adds complexity (order by index? order by column value?). Most useful for streaming queries where the total result may be too large to load into memory and sort.
```

Comment:
i need to read up on how `Arrow.RecordBatch` plays with things like `yield` paradigms & iteration over larger-than-RAM results ...

Reply:
it does not address that. Arrow is just an in-memory typing and serialization system.  A RecordBatch data structure - a general purpose group of named/typed arrays of equal length.  Any decisions about how to use it are outside the scope of the data structure.

the value in using it is that we now can utilize the fact that many other packages (eg, Pandas, numpy, etc) can import/export arrow types without any further transcoding (no copy, no serialization, etc - very efficient).  Using it here gives us a language- and environment-neutral means of defining schema & result "structure" without overly binding to specific packages (eg, whatever is used in Python/R/Julia/Go/...)

Reply:
@bkmartinjr what if the `Arrow.RecordBatch` is too big (perhaps far too big) for memory? i think we will need _some_ way of indicating partial results ...

Reply:
... even if it's an indication that people will need to call `read` multiple times in some kind of 'continue-from-last' paradigm (if that's the way we go) ...

Reply:
> what if the Arrow.RecordBatch is too big

See the previous bullet:  `- incremental/streaming queries`, which is the same as `partial results`

the RecordBatch is just a data structure holding the result of the query.   If the query/read is too big, it will fail.  So you are right that we need a way to `page` through the results.  I have not yet thought through how to best present that, which is why I listed it as an open (and critical) gap

Reply:
@bkmartinjr awesome! then we're both aware that we need to articulate (even at high level) something about how we want paging to feel like

Reply:
In thinking about this some more, I do not think the abstract SOMA spec should specify the _precise_ API for delayed and incremental reads.  Rather, it should say that it is available, and delegate to language-specific SOMA specs, as each environment will want to do it slightly differently.  I will add some proposed text to this effect.

Reply:
it needn't be _precise_ but i think we should specify _something_ or different impls will do wildly different things.

* does a larger-than-memory read return some kind of iterator which yields batches?
* does it return a single batch with a 'there is more' flag (like many restful web apis)?

Reply:
I am resolving this thread as the latest doc contains:
* use of Arrow for base type system and read/write ABI
* a more explicit discussion that clarifies the use of RecordBatch/Tensor as the ABI for a read/write "chunk", and defers the actual codification of a delayed IO iterator to the per-language bindings (ie, this doc just proposes that it is possible to do incremental reads, but doesn't specify how the R or Python bindings will express that).